packages/auth/__tests__/auth.int.jest.js

```diff
@@ -206,6 +206,75 @@ test('sign in with tokenType APP default tokenType', async () => {
   expect(resultSession2.signInResult.data.signIn.tokenType).toBe('EMAIL_TOKEN')
 })
 
+describe(`sign in with tokenType EMAIL_CODE`, () => {
+  const tokenType = 'EMAIL_CODE'
+  const MAX_VALID_TOKENS = 5
+
+  test('authorize within same session', async () => {
+    const resultSignIn = await signIn({ user: Users.Unverified, tokenType, skipAuthorization: true })
+    expect(resultSignIn.signInResult.data.signIn.phrase).toBeTruthy()
+    expect(resultSignIn.signInResult.errors).toBeFalsy()
+
+    const resultAuthorize = await authorizeSession({
+      email: Users.Unverified.email,
+      tokens: [{ type: tokenType, payload: resultSignIn.payload }],
+      apolloFetch: resultSignIn.apolloFetch
+    })
+
+    expect(resultAuthorize.data && resultAuthorize.data.authorizeSession).toBeTruthy()
+  })
+
+  test('denies authorization with different session', async () => {
+    const tokenType = 'EMAIL_CODE'
+
+    const resultSignIn = await signIn({ user: Users.Unverified, tokenType, skipAuthorization: true })
+    expect(resultSignIn.signInResult.data.signIn.phrase).toBeTruthy()
+    expect(resultSignIn.signInResult.errors).toBeFalsy()
+
+    const resultAuthorize = await authorizeSession({
+      email: Users.Unverified.email,
+      tokens: [{ type: tokenType, payload: resultSignIn.payload }],
+      apolloFetch: global.instance.createApolloFetch()
+    })
+
+    expect(resultAuthorize.data).toBeFalsy()
+  })
+
+  test('fails if amount of simultaniously valid tokens is exceeded', async () => {
+    const tokenType = 'EMAIL_CODE'
+
+    for (let i = 0; i < MAX_VALID_TOKENS; i++) {
+      await signIn({
+        user: Users.Unverified,
+        tokenType,
+        skipAuthorization: true
+      })
+    }
+
+    const resultSignIn = await signIn({ user: Users.Unverified, tokenType, skipAuthorization: true })
+    expect(resultSignIn.signInResult.errors).toBeTruthy()
```

Comment:
how is it possible to recover from this situtation? and could it be tested also?

Reply:
It is possible to recover by waiting for one or more tokens to expire.

This could be tested, too. :+1:

Reply:
Tests added. (Also some more.)